packages/core/src/browser/quick-input/quick-input-service.ts

```diff
@@ -0,0 +1,90 @@
+/********************************************************************************
+ * Copyright (c) 2021 SAP SE or an SAP affiliate company and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { injectable } from 'inversify';
+import { CancellationToken } from '../../common';
+import URI from '../../common/uri';
+import { QuickInputContribution } from './quick-input-contribution';
+
+export interface QuickPickItem {
+    type?: 'item' | 'separator';
+    id?: string;
+    label: string;
+    meta?: string;
+    ariaLabel?: string;
+    description?: string;
+    detail?: string;
+    resource?: URI;
+    iconClasses?: string[];
+    execute?: (quickPick: any, lookFor: string) => void;
+}
+export interface QuickPickValue<V> extends QuickPickItem {
+    value: V
+}
+
+@injectable()
+export class QuickInputService implements QuickInputContribution {
+    get backButton(): any { return; /* implemented by derived class */ }
+
+    open(filter: string): void { /* implemented by derived class */ }
```

Comment:
Please elaborate how would you separate `quickAccess` from `quickInput`?. `quickAccess` in property of `IQuickInputService`:

```
 export interface IQuickInputService {
        readonly _serviceBrand?: undefined;
        /**
         * Provides access to the back button in quick input.
         */
        readonly backButton?: IQuickInputButton;
        /**
         * Provides access to the quick access providers.
         */
        readonly quickAccess: IQuickAccessController;
```

Reply:
Even in VS Code, the `show(value = '', options?: IQuickAccessOptions): void` method is not on the IQuickInputService interface but on the `IQuickAccessController` interface. If you look at the implementation of the QuickInputService in VS Code, it never even _uses_ the QuickAccessController. On the contrary, the QuickAccessController uses the QuickInputService to show the quickpicks provided by the quick access providers.

So the real relationship is that the QuickAccessController `has a` QuickInputServices, not the other way around and there is certainly is no `is a` relationship.  I would suspect that the quick access controller is referenced off the QuickInputService out of convenience.

Reply:
So as far as the theia interfaces are concerned, the whole QuickAccessProvider stuff should not be connected to QuickInput, except through the fact that providers produce quick pick items. That the two might be connected in the implementation does not really matter, since we're using inversify to hook up the dependencies. IMO, we should have three interfaces:

```
QuickInputService {
    // equivalent to IQuickInputService
    createQuickPick(...)
    ...
}

QuickAccessRegistry {
   // equivalent of IQuickAccessRegistry, but with theia types, not monaco
   registerQuickAccessProvider(....);
   ...
}

QuickAccess {
    // equivalent of IQuickAccessController
    show(prefix, ...);
}
```
The should be usable without recourse to any monaco types and be completely independent of each other. Clients will get the monaco implementations injected, because that's what we bind by default.

Reply:
> Even in VS Code, the `show(value = '', options?: IQuickAccessOptions): void` method is not on the IQuickInputService interface but on the `IQuickAccessController` interface. If you look at the implementation of the QuickInputService in VS Code, it never even _uses_ the QuickAccessController. On the contrary, the QuickAccessController uses the QuickInputService to show the quickpicks provided by the quick access providers.
> 
> So the real relationship is that the QuickAccessController `has a` QuickInputService, not the other way around and there is certainly is no `is a` relationship. I would suspect that the quick access controller is referenced off the QuickInputService out of convenience.

Please check this [link](https://github.com/microsoft/vscode/blob/standalone/0.23.x/src/vs/platform/quickinput/browser/quickInput.ts) where you can see class `QuickInputService`  (which contains `quickAccess` prop) which I based `MonacoQuickInputService` on.

Reply:
> QuickAccessRegistry {
   // equivalent of IQuickAccessRegistry, but with theia types, not monaco
   registerQuickAccessProvider(....);
   ...
}

I do not agree about this - We are not registering a quick access provider in Theia - but in Monaco itself. Creating an interface `QuickAccessRegistry` similarly to monaco' `IQuickAccessRegistry` is like saying - we are managing in Theia our own Quick Access registry - while what we are trying to do here (in my solution) is to register out provider in Monaco's QuickAccessRegistry.

Therefore I created the `IQuickAccessDataService` which defines the methods required to be implemented by a "DataService" which will be used by provider - `getPicks` will be used to retrieve the data and 'registerQuickAccessProvider' for registering our QAP in Monaco's QA registry.

![image](https://user-images.githubusercontent.com/71069170/117962405-dabf5200-b327-11eb-9d53-4397585abe52.png)

Reply:
> is like saying - we are managing in Theia our own Quick Access registry

No, we're just saying that there is such a service and clients can use it. We then have an _implementation_ that uses the classes from VS Code (aka monaco).
The whole point of having the abstract layer of QuickInputService, etc. is to prevent clients to be tightly coupled to the VS Code implementation. The VS Code team gives absolutely no guarantee that API like `IQuickInputService` is stable in any way. That's why we have the abstractions in core: it allows us to have a layer where we can adapt to changes in the VS Code implementation without affecting the rest of our code. If tomorrow we decided to do our own implementation of the QuickInput UI, we could do that.

One of the reasons why it's hard to update monaco in Theia is that we let the monaco stuff leak into a lot of our codebase. There are some good reasons for that, but in general, we should strive to program against an abstraction, not VS Code implementation

Reply:
First, I'm all for doing the right solution. I'm discussing things with you for that goal.

Second:
> No, we're just saying that there is such a service and clients can use it. We then have an implementation that uses the classes from VS Code (aka monaco).

Again - there is a difference IMO between doing a "Theia implementation" of VS Code functionality and what is the purpose here for registering a QAP in monaco's QAP registry. 

The way you describe it, we are adding a `QuickAccessRegistry` class which is exposing a `registerQuickAccessProvider` method which for the user thinking of adding another QAP to Theia, could seem in face value as a class that has implemented a QAP registry internally and `registerQuickAccessProvider` is registering QAP inside itself - while in essence we are registering our QAP in monaco - i.e. we are not managing the different providers etc - monaco is managing it, we are just registering to it.

> The whole point of having the abstract layer of QuickInputService, etc. is to prevent clients to be tightly coupled to the VS Code implementation. 
> One of the reasons why it's hard to update monaco in Theia is that we let the monaco stuff leak into a lot of our codebase. There are some good reasons for that, but in general, we should strive to program against an abstraction, not VS Code implementation

Agree totally.

> The VS Code team gives absolutely no guarantee that API like IQuickInputService is stable in any way. That's why we have the abstractions in core: it allows us to have a layer where we can adapt to changes in the VS Code implementation without affecting the rest of our code. If tomorrow we decided to do our own implementation of the QuickInput UI, we could do that.

a) You can look at current PR as the result of Theia's previous implementation - meaning: Theia implemented a proprietary "file search" mechanism (onType) - and based itself on quickOpen (which is fine as that was what was available then) - but now, we had to refactor all usages of that `onType` mechanism - because VS Code refactored their end.
So you can have good intentions but their changes would require major refactoring. 

b) In addition, now, we have no choice but to base ourselves on VS Code `QuickInput` functionality - which entails a whole refactoring now which we could theoretically create a proprietary solution and not base ourselves on monaco quickInput but I'm guessing that isn't what you would want either - so I don't see another way but to base ourselves as closely as possible to VS Code implementation. As that would ensure we do the least of refactoring possible - IMO.